src/auth/client/mod.rs

```diff
@@ -78,48 +80,159 @@ pub enum LoginResponse {
 		redirect_address: (LuString33, u16),
 		/// Whether the account is in free trial mode.
 		is_ftp: bool,
+		/// Whether the account is connecting as a paid account for the first time.
+		just_upgraded_from_ftp: bool,
+		language: Language,
+		country_code: LuString3,
+		auth_logs: Vec<LogEntry>,
 	} = 1,
 	/// The login failed in an unusual way. More information can be found in the attached message.
 	CustomMessage(LuVarWString<u16>) = 5,
 	/// Username or password was incorrect.
 	InvalidUsernamePassword = 6,
 }
 
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct LogEntry {
+    pub type_: u32,
+    pub value: u32,
+    pub timestamp: u64,
+}
+
 impl<'a, W: LEWrite> Serialize<LE, W> for &'a LoginResponse
-	where       u8: Serialize<LE, W>,
-	           u16: Serialize<LE, W>,
-	           u32: Serialize<LE, W>,
-	      &'a bool: Serialize<LE, W>,
-	      &'a [u8]: Serialize<LE, W>,
-	   &'a LuString33: Serialize<LE, W>,
-	  &'a LuWString33: Serialize<LE, W>,
-	  &'a LuVarWString<u16>: Serialize<LE, W> {
-	fn serialize(self, writer: &mut W) -> Res<()>	{
-		let disc = unsafe { *(self as *const LoginResponse as *const u8) };
-		writer.write(disc)?;
-		match self {
-			LoginResponse::Ok { version, session_key, redirect_address, is_ftp } => {
-				writer.write(&[0; 264][..])?;
-				writer.write(version.0)?;
-				writer.write(version.1)?;
-				writer.write(version.2)?;
-				writer.write(session_key)?;
-				writer.write(&redirect_address.0)?;
-				writer.write(&[0; 33][..])?;
-				writer.write(redirect_address.1)?;
-				writer.write(&[0; 81][..])?;
-				writer.write(is_ftp)?;
-				writer.write(&[0; 10][..])?;
-			}
-			LoginResponse::CustomMessage(msg) => {
-				writer.write(&[0; 493][..])?;
-				writer.write(msg)?;
-			}
-			LoginResponse::InvalidUsernamePassword => {
-				writer.write(&[0; 495][..])?;
-			}
-		}
-		writer.write(4u32)?;
-		Ok(())
-	}
+where
+    u8: Serialize<LE, W>,
+    u16: Serialize<LE, W>,
+    u32: Serialize<LE, W>,
+    u64: Serialize<LE, W>,
+    &'a bool: Serialize<LE, W>,
+    &'a [u8]: Serialize<LE, W>,
+    &'a LuString33: Serialize<LE, W>,
+    &'a LuString33: Serialize<LE, W>,
+    &'a LuWString33: Serialize<LE, W>,
+    &'a LuString37: Serialize<LE, W>,
+    &'a LuString3: Serialize<LE, W>,
+    &'a Language: Serialize<LE, W>,
+    &'a LogEntry: Serialize<LE, W>,
+    &'a LuVarWString<u16>: Serialize<LE, W>,
+{
+    fn serialize(self, writer: &mut W) -> Res<()> {
+        let disc = unsafe { *(self as *const LoginResponse as *const u8) };
+        writer.write(disc)?;
+        match self {
+            LoginResponse::Ok {
+                version,
+                session_key,
+                redirect_address,
+                is_ftp,
+                just_upgraded_from_ftp,
+                language,
+                country_code,
+                auth_logs,
+            } => {
+                // event strings
+                writer.write(&b"Talk_Like_A_Pirate\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"[..])?;
```

Comment:
These strings are actually functional, even if not used in the captures, they can be used to unlock content gated in the cdclient EventGating table. Thus I'd prefer them to be implemented in case any server would use them, and for documentation. The other hardcoded things aren't strictly necessary, but while you're at it, it would be nice to have them I guess, just so we don't need to add them in the future. Now that the rest of the deserialization is there, this should only be a few line changes.

Reply:
> These strings are actually functional, even if not used in the captures, they can be used to unlock content gated in the cdclient EventGating table. 

My bad, forgot about that. I've added all of the other stuff as well.